src/CommunityToolkit.Maui/Views/Popup/PopupPage.shared.cs

```diff
@@ -0,0 +1,185 @@
+	{
+		ArgumentNullException.ThrowIfNull(popup);
+		ArgumentNullException.ThrowIfNull(popupOptions);
+
+		this.popup = popup;
+		this.popupOptions = popupOptions;
+
+		// Only set the content if parent constructor hasn't set the content already; don't override content if it already exists
+		base.Content ??= new PopupPageLayout(popup, popupOptions);
+
+		tapOutsideOfPopupCommand = new Command(async () =>
+		{
+			popupOptions.OnTappingOutsideOfPopup?.Invoke();
+			await Close(new PopupResult(true));
+		}, () => popupOptions.CanBeDismissedByTappingOutsideOfPopup);
+
+		Content.GestureRecognizers.Add(new TapGestureRecognizer { Command = tapOutsideOfPopupCommand });
+
+		if (popupOptions is BindableObject bindablePopupOptions)
+		{
+			bindablePopupOptions.PropertyChanged += HandlePopupPropertyChanged;
+		}
+
+		this.SetBinding(BindingContextProperty, static (Popup x) => x.BindingContext, source: popup, mode: BindingMode.OneWay);
+		this.SetBinding(BackgroundColorProperty, static (IPopupOptions options) => options.PageOverlayColor, source: popupOptions, mode: BindingMode.OneWay);
+
+		Shell.SetPresentationMode(this, PresentationMode.ModalNotAnimated);
+		On<iOS>().SetModalPresentationStyle(UIModalPresentationStyle.OverFullScreen);
+	}
+
+	public event EventHandler<IPopupResult> PopupClosed
+	{
+		add => popupClosedEventManager.AddEventHandler(value);
+		remove => popupClosedEventManager.RemoveEventHandler(value);
+	}
+
+	// Prevent Content from being set by external class
+	// Casts `PopupPage.Content` to return typeof(PopupPageLayout)
+	internal new PopupPageLayout Content => (PopupPageLayout)base.Content;
+
+	public async Task Close(PopupResult result, CancellationToken token = default)
+	{
+		token.ThrowIfCancellationRequested();
+
+		var popupPageToClose = Navigation.ModalStack.OfType<PopupPage>().LastOrDefault(popupPage => popupPage.Content == Content);
+
+		if (popupPageToClose is null)
+		{
+			throw new InvalidOperationException($"Unable to close popup: could not locate {nameof(PopupPage)}. If using a custom implementation of {nameof(Popup)}, override the {nameof(Close)} method");
+		}
+
+		if (Navigation.ModalStack[^1] is Microsoft.Maui.Controls.Page currentVisibleModalPage
+			&& currentVisibleModalPage != popupPageToClose)
+		{
+			throw new InvalidOperationException($"Unable to close Popup because it is blocked by the Modal Page {currentVisibleModalPage.GetType().FullName}. Please call `{nameof(Navigation)}.{nameof(Navigation.PopModalAsync)}()` to first remove {currentVisibleModalPage.GetType().FullName} from the {nameof(Navigation.ModalStack)}");
+		}
+
+		await Navigation.PopModalAsync(false).WaitAsync(token);
+
+		popupClosedEventManager.HandleEvent(this, result, nameof(PopupClosed));
+	}
+
+	// Prevent the Android Back Button from dismissing the Popup if CanBeDismissedByTappingOutsideOfPopup is true
+	protected override bool OnBackButtonPressed()
+	{
+		if (popupOptions.CanBeDismissedByTappingOutsideOfPopup)
```

Comment:
Ignore this if we choose the `PopupOptions` to not be a `BindableObject`

this doesn't count for the `PropertyChanged`.

Reply:
I'm not sure what you mean. Could you elaborate?

Reply:
Ignore this one.